crates/bevy_reflect/src/type_registry.rs

```diff
@@ -100,24 +107,57 @@ impl TypeRegistry {
         registry
     }
 
-    /// Registers the type `T`, adding reflect data as specified in the [`Reflect`] derive:
+    /// Attempts to register the type `T` if it has not yet been registered already.
+    ///
+    /// If the registration for type `T` already exists, it will not be registered again.
+    /// To register the type, overwriting any existing registration, use [register](Self::register) instead.
+    ///
+    /// Additionally, this will add the reflect [data](TypeData) as specified in the [`Reflect`] derive:
     /// ```ignore (Neither bevy_ecs nor serde "derive" are available.)
-    /// #[derive(Component, serde::Serialize, serde::Deserialize, Reflect)]
+    /// #[derive(Reflect)]
     /// #[reflect(Component, Serialize, Deserialize)] // will register ReflectComponent, ReflectSerialize, ReflectDeserialize
+    /// struct Foo;
     /// ```
     pub fn register<T>(&mut self)
     where
         T: GetTypeRegistration,
     {
-        self.add_registration(T::get_type_registration());
+        if self.contains(TypeId::of::<T>()) {
+            return;
+        }
+
+        self.overwrite_registration(T::get_type_registration());
+        T::register_type_dependencies(self);
     }
 
-    /// Registers the type described by `registration`.
-    pub fn add_registration(&mut self, registration: TypeRegistration) {
-        if self.registrations.contains_key(&registration.type_id()) {
-            return;
+    /// Attempts to register the type described by `registration`.
+    ///
+    /// If the registration for the type already exists, it will not be registered again.
+    ///
+    /// To forcibly register the type, overwriting any existing registration, use the
+    /// [`overwrite_registration`](Self::overwrite_registration) method instead.
+    ///
+    /// Returns `None` if the registration was successfully added,
+    /// or `Some(existing)` if it already exists.
+    pub fn add_registration(
+        &mut self,
+        registration: TypeRegistration,
+    ) -> Option<&TypeRegistration> {
+        if self.contains(registration.type_id()) {
+            self.get(registration.type_id())
+        } else {
+            self.overwrite_registration(registration);
+            None
         }
```

Comment:
@alice-i-cecile @soqb For some reason I was having a lot of issues getting the borrow checker to be happy with me haha. I don't love that it's adding a second lookup, but it's the best I could come up with right now. Let me know if there's a smarter way fo doing this lol